### PR TITLE
Not render label when field is none + cache invoice report

### PR DESCRIPTION
**Pull request by raimonesteve**
#043353

### CODE REVIEW

File: invoice/invoice.py [hunk 1]
```diff
@@ -45,3 +47,36 @@ def get_sorted_key(self, name):
...
+        if len(ids) == 1:
+            # Re-instantiate because records are TranslateModel
+            invoice, = Invoice.browse(ids)
+            if invoice.invoice_report_cache:
+                return (
+                    invoice.invoice_report_format,
+                    bytes(invoice.invoice_report_cache))
+            else:
+
+                if invoice.state in {'posted', 'paid'} and invoice.type == 'out':
```
Comment (albertca) [inline invoice/invoice.py:76]: This could become an "elif".

File: invoice/invoice.py [hunk 2]
```diff
@@ -45,3 +47,36 @@ def get_sorted_key(self, name):
...
+    def execute(cls, ids, data):
+        pool = Pool()
+        Invoice = pool.get('account.invoice')
+
+        result = super(InvoiceReport, cls).execute(ids, data)
+
+        if len(ids) == 1:
+            # Re-instantiate because records are TranslateModel
+            invoice, = Invoice.browse(ids)
+            if invoice.invoice_report_cache:
```
Comment (albertca) [inline invoice/invoice.py:70]: Cache should be checked before calling super() because if there's already a cache we do not wan to render the report again, which is what super will do.

File: invoice/invoice.py [hunk 3]
```diff
@@ -45,3 +47,36 @@ def get_sorted_key(self, name):
...
+
+    @classmethod
+    def execute(cls, ids, data):
+        pool = Pool()
+        Invoice = pool.get('account.invoice')
+
+        result = super(InvoiceReport, cls).execute(ids, data)
+
+        if len(ids) == 1:
+            # Re-instantiate because records are TranslateModel
```
Comment (albertca) [inline invoice/invoice.py:68]: I think this comment does not apply here.

File: engine.py [hunk 4]
```diff
@@ -554,6 +554,9 @@ def label(cls, model, field=None, lang=None):
         if not lang:
             lang = Transaction().language
 
+        if not model:
```
Comment (albertca) [inline engine.py:557]: Why is this necessary. Why would anyone ask for a label of an empty model?

The only possibility I see is that the model may not exist and thus line 562 would crash:

model, = Model.search([('model', '=', model)])

Then we could return model variable instead of model.name.

**Review by albertca — CHANGES_REQUESTED** [review]

File: invoice/invoice.py (hunk 3 above)
Comment (raimonesteve) [reply invoice/invoice.py:68]: copy from https://github.com/tryton/account_invoice/blob/develop/invoice.py#L2666

File: engine.py (hunk 4 above)
Comment (raimonesteve) [reply engine.py:557]: ... crash when try to get label in a model that not exist (header report) because "model" is None.

I don't remember exact the label that crash.... :( maybe is related to signature template [1] ? 

Example:

{{ label(None, 'reference') }}

[1] https://github.com/trytonspain/trytond-html_report/blob/master/html.py#L140

**Review by albertca — APPROVED** [review]
